### ISSUE TITLE
Cmamba cannot be trained

### ISSUE THREAD

**Issue by CszYihen** [MShahabSepehri/CryptoMamba#18]
When I trained Cmamba, all the loss indicators returned nan。However, the training of the GRU model was normal, and it was all conducted using the same dataset.

**Comment by MShahabSepehri** [comment]
I need more details about the training. Are you using our configurations?
We have tested the current version on Colab, and it is working.

**Comment by CszYihen** [comment]
These are my training orders.
```sh
python scripts/training.py --config cmamba_nv --save_checkpoints 
```
This is his output.
```
Seed set to 23
['Timestamp', 'Open', 'High', 'Low', 'Close']
['Timestamp', 'Open', 'High', 'Low', 'Close']
['Timestamp', 'Open', 'High', 'Low', 'Close']
2025-05-28 15:34:46.387955: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2025-05-28 15:34:47.031675: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
[rank: 0] Seed set to 23
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/1
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 1 processes
----------------------------------------------------------------------------------------------------

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]

  | Name  | Type                        | Params
------------------------------------------------------
0 | mse   | MSELoss                     | 0     
1 | l1    | L1Loss                      | 0     
2 | mape  | MeanAbsolutePercentageError | 0     
3 | model | CMamba                      | 136 K 
------------------------------------------------------
136 K     Trainable params
0         Non-trainable params
136 K     Total params
0.548     Total estimated model params size (MB)
Sanity Checking: |                                                                                                                                      | 0/? [00:00<?, ?it/s]350 data points loaded as val split.
1446 data points loaded as train split.                                                                                                                                       
Epoch 0: 100%|███████████████████████████████████████████████████| 46/46 [00:01<00:00, 24.15it/s, v_num=4, train/rmse=nan.0, train/mape=nan.0, val/rmse=nan.0, val/mape=nan.0]Epoch 0, global step 46: 'val/rmse' reached inf (best inf), saving model to 'logs/CMamba/version_4/checkpoints/epoch0-val-rmsenan.ckpt' as top 1                              
Epoch 1: 100%|███████████████████████████████████████████████████| 46/46 [00:02<00:00, 22.62it/s, v_num=4, train/rmse=nan.0, train/mape=nan.0, val/rmse=nan.0, val/mape=nan.0]Epoch 1, global step 92: 'val/rmse' was not in top 1                                                                                                                          
Epoch 2: 100%|███████████████████████████████████████████████████| 46/46 [00:02<00:00, 22.59it/s, v_num=4, train/rmse=nan.0, train/mape=nan.0, val/rmse=nan.0, val/mape=nan.0]Epoch 2, global step 138: 'val/rmse' was not in top 1                                                                                                                         
Epoch 3: 100%|███████████████████████████████████████████████████| 46/46 [00:02<00:00, 22.50it/s, v_num=4, train/rmse=nan.0, train/mape=nan.0, val/rmse=nan.0, val/mape=nan.0]Epoch 3, global step 184: 'val/rmse' was not in top 1                                                                                                                         
Epoch 4:  83%|██████████████████████████████████████████▏        | 38/46 [00:01<00:00, 25.75it/s, v_num=4, train/rmse=nan.0, train/mape=nan.0, val/rmse=nan.0, val/mape=nan.0]^C/home/yang/data/anaconda3/envs/mamba/lib/python3.10/site-packages/pytorch_lightning/trainer/call.py:54: Detected KeyboardInterrupt, attempting graceful shutdown...
[rank: 0] Seed set to 23
Restoring states from the checkpoint path at logs/CMamba/version_4/checkpoints/epoch0-val-rmsenan.ckpt
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Loaded model weights from the checkpoint at logs/CMamba/version_4/checkpoints/epoch0-val-rmsenan.ckpt
350 data points loaded as test split.
Testing DataLoader 0: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 11/11 [00:00<00:00, 101.37it/s]
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃        Test metric        ┃       DataLoader 0        ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│         test/mae          │            nan            │
│         test/mape         │            nan            │
│         test/mse          │            nan            │
│         test/rmse         │            nan            │
└───────────────────────────┴───────────────────────────┘
```
The output of his evaluation indicators is all NAN。

**Comment by 156ak** [comment]
Hi @MShahabSepehri and the CryptoMamba team,
First of all, congratulations on the publication and presentation of your work at ICBC and ICLR! This is a fascinating project, and using a Mamba-based SSM for crypto price prediction is a very innovative approach. Thank you for making the code and data publicly available.
I was exploring the repository and realized that providing a Dockerfile would be a great addition. It would significantly simplify the deployment process for many users, ensuring a consistent and reproducible environment across different machines. This could help lower the entry barrier for new contributors and users who want to test or build upon your work.
Thanks again for your excellent work!
Best regards.

**Comment by MShahabSepehri** [comment]
Hi @156ak. 
Thank you for your support and interest. 
This is a great idea, and having a Dockerfile can improve usability and reproducability. Unfortunately, we currently don't have the bandwidth to implement it ourselves, but we would be happy to support any community contributions. If you or anyone else in the community is interested in creating one, please feel free to submit a pull request, and we’d be glad to review and integrate it.

**Comment by aaronprivate95-glitch** [comment]
NaN loss usually boils down to one of a few repeat offenders in time-series forecasting codepaths — here are the fastest checks/fixes (especially given you run GRU fine on the same data):

## 1) Check targets/inputs for NaN/Inf after preprocessing
Even if raw OHLC is clean, transforms can introduce Inf:
- log / pct change on 0 or negative values
- standardization with std=0 on a constant column
- rolling indicators with insufficient warmup

Quick debug:
- print min/max + count_nan/count_inf right before the model forward.

## 2) MAPE can explode / become NaN when y is 0 (or very close)
Your logs show rmse+mape are NaN immediately.
- If the code computes MAPE as |(y - yhat) / y|, then any y==0 will produce Inf/NaN.
Fix:
- use epsilon: denom = max(|y|, eps)
- or switch to sMAPE / MAE for training-time reporting.

## 3) Learning rate / grad explosion
Even with small models, SSM-style blocks can blow up with too high LR.
Mitigations:
- lower LR 10x and see if NaN disappears
- enable gradient clipping (e.g., 0.5–1.0)
- disable mixed precision if enabled (fp16/bf16 instability)

## 4) Normalization mismatch
If GRU is trained on normalized series but CMamba config expects different scaling (e.g., z-score vs minmax), NaNs can appear in the first few steps.
- verify config cmamba_nv normalization section matches the dataset pipeline.

If you share: (a) a small sample batch after preprocessing, (b) the exact config values for LR/precision/normalization, it’s usually possible to pinpoint the exact source within 1–2 iterations.

Hands-on debugging help / model audit: https://gist.github.com/aaronprivate95-glitch/8014d76a8675751f4f67da32b668bc0a
